File: solution/00/05/sol_dynamic.cc
#include <bits/stdc++.h>

using namespace std;

static const int _ = []()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution
{
public:
    static string longestPalindrome(const string& s)
    {
        const auto n = s.size();
        if (n == 0) return "";
        vector dp(n, vector(n, false));

        int start = 0;
        int max_len = 1;

        for (int i = 0; i < n; ++i)
        {
            dp[i][i] = true;
        }

        for (int i = 0; i < n - 1; ++i)
        {
            if (s[i] == s[i + 1])
            {
                dp[i][i + 1] = true;
                start = i;
                max_len = 2;
            }
        }

        for (int length = 3; length <= n; ++length)
        {
            for (int i = 0; i <= n - length; ++i)
            {
                int j = i + length - 1;
                if (s[i] == s[j] && dp[i + 1][j - 1])
                {
                    dp[i][j] = true;
                    if (length > max_len)
                    {
                        start = i;
                        max_len = length;
                    }
                }
            }
        }
        return s.substr(start, max_len);
    }
};

int main()
{
    const string s = "babad";
    cout << Solution::longestPalindrome(s) << endl;
    return 0;
}

File: solution/00/05/sol_dynamic.py
class Solution:
    def longestPalindrome(self, s: str) -> str:  # noqa
        n = len(s)
        if n == 0:
            return ""
        if n == 1:
            return s
        dp = [[False] * n for _ in range(n)]
        for i in range(n):
            dp[i][i] = True
        max_len = 1
        start = 0
        for j in range(1, n):
            for i in range(j):
                if s[i] == s[j]:
                    if j - i < 3:
                        dp[i][j] = True
                    else:
                        dp[i][j] = dp[i + 1][j - 1]
                else:
                    dp[i][j] = False
                if dp[i][j]:
                    cur_len = j - i + 1
                    if cur_len > max_len:
                        max_len = cur_len
                        start = i
        return s[start:start + max_len]


def main():
    s = Solution()
    assert (result := s.longestPalindrome("babad")) == "bab", f"unexpected {result=}"
    assert (result := s.longestPalindrome("cbbd")) == "bb", f"unexpected {result=}"
    assert (result := s.longestPalindrome("a")) == "a", f"unexpected {result=}"

    print("PASSED")


if __name__ == '__main__':
    main()
